src/compiler/checker.ts

```diff
@@ -10547,62 +10547,102 @@ module ts {
             return false;
         }
 
-        function getSymbolsInScope(location: Node, meaning: SymbolFlags): Symbol[] {
+        function getSymbolsInScope(location: Node, meaning: SymbolFlags, predicate?: (symbol: Symbol) => boolean): Symbol[] {
             let symbols: SymbolTable = {};
             let memberFlags: NodeFlags = 0;
-            function copySymbol(symbol: Symbol, meaning: SymbolFlags) {
+
+            if (isInsideWithStatementBody(location)) {
+                // We cannot answer semantic questions within a with block, do not proceed any further
+                return [];
+            }
+
+            populateSymbols();
+
+            return mapToArray(symbols);
+
+            function populateSymbols() {
+                while (location) {
+                    if (location.locals && !isGlobalSourceFile(location)) {
+                        if (copySymbols(location.locals, meaning)) {
+                            return;
+                        }
+                    }
+                    switch (location.kind) {
+                        case SyntaxKind.SourceFile:
+                            if (!isExternalModule(<SourceFile>location)) {
+                                break;
+                            }
+                        case SyntaxKind.ModuleDeclaration:
+                            if (copySymbols(getSymbolOfNode(location).exports, meaning & SymbolFlags.ModuleMember)) {
+                                return;
+                            }
+                            break;
+                        case SyntaxKind.EnumDeclaration:
+                            if (copySymbols(getSymbolOfNode(location).exports, meaning & SymbolFlags.EnumMember)) {
+                                return;
+                            }
+                            break;
+                        case SyntaxKind.ClassDeclaration:
+                        case SyntaxKind.InterfaceDeclaration:
+                            if (!(memberFlags & NodeFlags.Static)) {
+                                if (copySymbols(getSymbolOfNode(location).members, meaning & SymbolFlags.Type)) {
+                                    return;
+                                }
+                            }
+                            break;
+                        case SyntaxKind.FunctionExpression:
+                            if ((<FunctionExpression>location).name) {
+                                if (copySymbol(location.symbol, meaning)) {
+                                    return;
+                                }
+                            }
+                            break;
+                    }
+                    memberFlags = location.flags;
+                    location = location.parent;
+                }
+                if (copySymbols(globals, meaning)) {
+                    return;
+                }

```

Comment:
Is this `if` necessary beyond the desire for parallelism?

Note that I don't mean parallelism in the computing context, I mean it as a literary device in which components of a text are constructed in a similar manner.